File: 0121-best-time-to-buy-and-sell-stock/solution.go
func maxProfit(prices []int) int {
    // two pointer
    maxProfit := 0
    left, right := 0, 1

    for right < len(prices) {
        // check if profitable to buy at left, and sell at right
        if prices[left] < prices[right] {
            if prices[right] - prices[left] > maxProfit {
                maxProfit = prices[right] - prices[left]
            }
        } else {
            // set left to right, because right has the lowest price
            left = right
        }
        
        right++
    }

    return maxProfit
}

File: 0121-best-time-to-buy-and-sell-stock/solution.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let min = 1_00_001;
    let profit = 0;
    
    for (let i = 0; i < prices.length; i++) {
        let price = prices[i];
        if (price < min) {
            min = price;
        }

        let profitTemp = price - min;
        if (profitTemp > profit) {
            profit = profitTemp;
        }
    }

    return profit;
};
